index.js

```diff
@@ -413,6 +413,105 @@ NYC.prototype.report = function () {
   }
 }
 
+// XXX(@isaacs) Index generation should move to istanbul-lib-processinfo
+NYC.prototype.writeProcessIndex = function () {
+  const dir = this.processInfoDirectory()
+  const pidToUid = new Map()
+  const infoByUid = new Map()
+  const eidToUid = new Map()
+  const infos = fs.readdirSync(dir).filter(f => f !== 'index.json').map(f => {
+    try {
+      const info = JSON.parse(fs.readFileSync(path.resolve(dir, f), 'utf-8'))
+      // on thiis first read pass, also map the pids to uuids
+      info.uuid = path.basename(f, '.json')
+      pidToUid.set(info.uuid, info.pid)
+      pidToUid.set(info.pid, info.uuid)
+      infoByUid.set(info.uuid, info)
+      if (info.externalId) {
+        eidToUid.set(info.externalId, info.uuid)
+      }
+      return info
+    } catch (er) {
+      return null
+    }
+  }).filter(Boolean)
+
+  // create all the parent-child links and write back the updated info
+  const needsUpdate = new Set()
+  infos.forEach(info => {
+    if (info.ppid && info.ppid !== '0' && !info.parent) {
+      info.parent = pidToUid.get(info.ppid)
+      needsUpdate.add(info)
+    }
+    if (info.parent) {
+      const parentInfo = infoByUid.get(info.parent)
+      if (parentInfo.children.indexOf(info.uuid) === -1) {
+        parentInfo.children.push(info.uuid)
+        needsUpdate.add(parentInfo)
+      }
+    }
+  })
+
+  // figure out which files were touched by each process.
+  const files = infos.reduce((files, info) => {
+    if (!info.files) {
+      try {
+        info.files = Object.keys(JSON.parse(fs.readFileSync(
+          path.resolve(this.tempDirectory(), info.coverageFilename),
+          'utf-8'
+        )))
+      } catch (er) {
+        return files
+      }
+      needsUpdate.add(info)
+    }
+    info.files.forEach(f => {
+      files[f] = files[f] || []
+      files[f].push(info.uuid)
+    })
+    return files
+  }, {})
+
+  // build the actual index!
+  const index = infos.reduce((index, info) => {
+    index.processes[info.uuid] = {}
+    if (info.parent) {
+      index.processes[info.uuid].parent = info.parent
+    }
+    if (info.externalId) {
+      index.processes[info.uuid].externalId = info.externalId
+      index.externalIds[info.externalId] = {
```

Comment:
What if someone accidentally uses the same externalId twice?  Not suggesting that needs to be supported, IMO it's probably an error.  I wouldn't object to a simple error:
```js
throw new Error(`External ID ${info.externalId} used by multiple processes`)
```

Reply:
Isn't that sort of the point, though?  Fairly soon, I'd like to be able to re-run a test with a matching externalId, and have nyc know to clean up _just_ the coverage and processinfo associated with that test.  I suppose while that is still pending, it's going to produce a weird result, but also, we're not actually using the index for anything until then either.

Reply:
Hmmm my concern is for situations where you have 2 different tests using the same externalId.  How can we catch that?

Reply:
I think, in a subsequent update, we "catch" that by seeing that an externalId is already already in the index at the start of the process, and clearing away its coverage and processinfo entries, so that when the index is regenerated, it's no longer present.

I see what you're saying, though, that you might have something like this:

```js
spawn(node, 'foo.js', { env: { NYC_PROCESSINFO_EXTERNAL_ID: 'foo' } })
spawn(node, 'bar.js', { env: { NYC_PROCESSINFO_EXTERNAL_ID: 'foo' } })
```

right next to each other.

In that case, yes, it's user-error, and now that I think about it, when we do automatically clear out the external id'd files on subsequent runs, it'll never happen _except_ when there's a race in the same process group, if we do it right, which would make a throw here appropriate.

The `--watch` pseudocode then looks like:

- run `EID=test/foo.js node test/foo.js`
- run `EID=test/bar.js node test/bar.js`
- Generate index (test/foo.js covers lib/foo.js, test/bar.js covers lib/bar.js)
- detect change in `lib/foo.js`, consult index.json for which procs to re-run
- run `EID=test/foo.js node test/foo.js`
  - nyc finds `test/foo.js` externalId in index.json, remove all coverage and processinfo data (NB: index.json now out of sync)
  - nyc runs `node test/foo.js`, generates coverage and processinfo files
- Generate index (still only one `EID=test/foo.js` present)

Care would have to be taken that if there's _another_ change to lib/foo.js while the test is running, we don't run the next run until the first is done.  (Or kill it or something.)  But a lot of tests will need that anyway, because of fixtures and the like, so it's not a _new_ concern for tap/ava/whoever.

Ok, yes, throw here.  Will do :)